### DISCUSSION TITLE
Deprecation of features

### DISCUSSION THREAD

**Discussion by jcorporation** [jcorporation/myMPD#356] [General]
Following features are considered deprecated and are removed in the 7.x versions.

- System commands: use the scripting feature instead
- Bookmarks: use home icons instead
- Chroot - use docker instead
- readonly - use a overlayfs or tmpfs instead
- Option to disable coverimages - everyone wants coverimages ;)
- Mixrampdb settings - does someone use this feature of MPD?
- Scrobbler integration - will be replaced with scrobbler implementation in lua
 - Webdav support - mongoose 7 does not support webdav anymore
- Configureable mpd stream uri - local playback feature now uses the integrated reverse proxy for streaming

**Comment by GK-774** [comment]
"System commands: use the scripting feature instead" - that would be unfortunate. It´s any easy way to provide a mechanism, for e.g. cleanly shutting down their music server, to terminal agnostic people.

**Reply by jcorporation** [reply]
The same can be achieved with a simple script. Lua has a os.execute function.

**Comment by GK-774** [comment]
Well, that would need to run as root, no ? Anyway, on Centos 7 Lua is to old for myMPD for example, so all this does is replacing a simple and working solution by a new dependency and a more complex solution, or as the lua documentation says: "The os.execute function is powerful, but it is also highly system dependent.". (The same thing applies to "Chroot - use docker instead" BTW.)

**Comment by jcorporation** [comment]
> cleanly shutting down their music server

this was the initial intention for the system commands feature. This feature has some drawbacks:

- it blocks the handling myMPD thread
- it can not capture the output of the command

Both is fixed in the scripting feature and you can do even more with lua scripting. E.g. accessing all the myMPD API commands. Using os.execute should by straight forward on all linux distributions and works also with sudo as the system commands. 

Why do you run not an actual version of centos? I think mpd itself is rather outdated.

The scripting feature was not developed to only replace the system commands, it can do much more, why should I maintain an old code path that is obsolete?

**Comment by GK-774** [comment]
"you can do even more with lua scripting" thats certainly something which those to whom a shutdown command via a webinterface is targeted wont do. Anyway, so this means for one simple shell command like "sudo shutdown -h now" I will need to have install a scripting language and if it´s not recent enough update the OS ? Strikes me as rather unelegant. But ok, then this is so.

**Reply by jcorporation** [reply]
I know what you mean, but a recent os is not only good for myMPD. 

The problem is, that I have no statistics about feature usage. The feedback from myMPD users is rather minimal.

If more users prefer keeping system commands, I will refactor it.

But how I would get enough feedback? Displaying deprecation warnings in the user interface?

The „What functions do you use?“ thread was not useful.

**Reply by Emilot** [reply]
I agree with GK-774 in this JC.

System commands are not only useful, it is a main core issue. I do not want to install unnecessary things just for reboot or shutdown my RPi.

I do not want to draw more resources that is absolutely vital to my system. Clean, minimal installation is an audiophile installation.

**Reply by jcorporation** [reply]
I think linking lua do not bloat myMPD in any substantial way. Lua is not big evil resource  hungry daemon, it is a small embedded scripting language.

Removing old and outdated code also frees up resources and makes the code maintenance simplier.

**Reply by Emilot** [reply]
Maybe you can leave it both and the decision being made by the client?

**Reply by jcorporation** [reply]
I want streamline the code base and do not want two so similar features.

**Reply by Emilot** [reply]
No prob JC, you are right :) 

Just please include one or two coding examples when you deprecate system commands.

**Reply by jcorporation** [reply]
Scripting is already documented with some examples:
https://github.com/jcorporation/myMPD/wiki/Scripting

**Comment by jdr85** [comment]
Although it's a pity that system commands have been or are to be deprecated (I used them also for just shutting down the system properly and easily), it was really simple to implement Lua based shutdown and it does not feel bloated at all. 

@jcorporation Jürgen, have system commands really been deprecated yet or will this happen in the near future? I've just upgraded to 7.0.2 and my old syscmds based shutdown is still in the menu and it even works which surprises me...

Regarding a recent OS: Best decision ever (besides myMPD of course) was to use [Arch Linux ARM](https://archlinuxarm.org/) on my music player Pi 1B as it is fast, does not have any issues with outdated software and is definitely more than stable enough to use it as a music player.

How it worked for me:
- Install necessary Lua package if not already done (this was just called "lua" on Arch)
- ./build.sh releaseinstall
- Modify mympd.conf as described [here](https://github.com/jcorporation/myMPD/wiki/Scripting) although it should probably be scripting = true instead of scripting = false under [mympd]; I just load the library "os".
- Create /var/lib/mympd/scripts/Ausschalten.lua (owner mpd:mpd, permissions 0700), content see below
- Of course /etc/sudoers must be configured properly
- Works!

Content of Ausschalten.lua:
`-- {"order":1,"arguments":[]}`
`os.execute("sudo /sbin/poweroff")`

**Comment by jcorporation** [comment]
In the v8.0 branch are all deprecated features removed. In the starke Branche only WebDAV was removed. Such changes are mentioned in the release notes.
